testresultparser/parser.go

```diff
@@ -0,0 +1,180 @@
+package testresultparser
+
+import (
+	"encoding/xml"
+	"io/ioutil"
+	"os"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// MVNFailure captures failure noted in Test Case in XML Report
+type MVNFailure struct {
+	FailureType string `xml:"type,attr"`
+	Message     string `xml:",chardata"`
+}
+
+// TestCase records test case info in XML Report
+type TestCase struct {
+	Name string     `xml:"name,attr"`
+	F    MVNFailure `xml:"failure"` // struct for Failure Details
+}
+
+// TestResultXML is struct for XML Test Report
+type TestResultXML struct {
+	TestSuite string     `xml:"name,attr"`
+	Tests     int        `xml:"tests,attr"`
+	Failures  int        `xml:"failures,attr"`
+	Errors    int        `xml:"errors,attr"`
+	Skipped   int        `xml:"skipped,attr"`
+	Time      string     `xml:"time,attr"`
+	T         []TestCase `xml:"testcase"` // struct for each Test Case
+}
+
+// failsafeReportXML is struct for Failafe Reports
+type failsafeReportXML struct {
+	Result         string `xml:"result,attr"`
+	TimeOut        string `xml:"timeout,attr"`
+	Tests          int    `xml:"completed"`
+	Failures       int    `xml:"failures"`
+	Errors         int    `xml:"errors"`
+	Skipped        int    `xml:"skipped"`
+	FailureMessage string `xml:"failureMessage"`
+}
+
+// Failure struct is the part of the Result struct for the output of the Parser
+type Failure struct {
+	TestCase string
+	Type     string
+	Message  string
+}
+
+// Result is the output strcut of the Parser
+type Result struct {
+	TestSuite string
+	Tests     int
+	Failures  int
+	Errors    int
+	Skipped   int
+	Time      string
+	Failure   []Failure
+}
+
+// convertResultFromFailsafe is a helper function
+// to convert struct to desired output struct
+func convertResultFromFailsafe(f failsafeReportXML) Result {
+	result := Result{
+		TestSuite: f.Result,
+		Tests:     f.Tests,
+		Failures:  f.Failures,
+		Errors:    f.Errors,
+		Skipped:   f.Skipped,
+		Time:      f.TimeOut,
+	}
+	result.Failure = append(result.Failure,
+		Failure{TestCase: "",
+			Type:    "",
+			Message: f.FailureMessage,
+		})
+	return result
+}
+
+// convertResultFromXML is a helper function
+// to convert struct to desired output struct
+func convertResultFromXML(t TestResultXML) Result {
+	result := Result{
+		TestSuite: t.TestSuite,
+		Tests:     t.Tests,
+		Failures:  t.Failures,
+		Errors:    t.Errors,
+		Skipped:   t.Skipped,
+		Time:      t.Time}
+
+	for _, test := range t.T {
+		if test.F.FailureType != "" {
+			result.Failure = append(result.Failure,
+				Failure{TestCase: test.Name,
+					Type:    test.F.FailureType,
+					Message: test.F.Message,
+				})
+		}
+	}
+	return result
+}
+
+// Parse method for Text Report parsing
+func (r *Result) Parse(f []byte) *Result {
```

Comment:
I think we are mixing data structure with behaviour here. I would rather have `Result` or even `TestResult` as a unified type for all the supported reports and provide concrete types with `Parse` implementation and intermediate datastructures. So instead of having `TestResultXML`, `failsafeReportXML`, `MVNFailure` and so on I would rather think about having `MavenParser` which internally verifies if the given file is failsafe or surefire, do all the parsing job and return the `Result`. The same with `GradleParser`